src/units/udisplay.cpp

```diff
@@ -640,33 +648,43 @@ void unit_attack(display * disp, game_board & board,
 	unit_animator animator;
 
 	animator.add_animation(attacker.shared_from_this(), "attack", att->get_location(), def->get_location(), damage, true, text_2,
-		(drain_amount >= 0) ? color_t(0, 255, 0) : color_t(255, 0, 0), hit_type, attack.shared_from_this(),
+		(drain_amount >= 0) ? color_t(0, 255, 0) : color_t(255, 0, 0), hit_type, weapon,
 		secondary_attack, swing);
 
 	// note that we take an anim from the real unit, we'll use it later
 	const unit_animation* defender_anim = def->anim_comp().choose_animation(*disp, def->get_location(), "defend",
-		att->get_location(), damage, hit_type, attack.shared_from_this(), secondary_attack, swing);
+		att->get_location(), damage, hit_type, weapon, secondary_attack, swing);
 
 	animator.add_animation(defender.shared_from_this(), defender_anim, def->get_location(), true, text, {255, 0, 0});
 
-	for(const unit_ability& ability : attacker.get_abilities_weapons("leadership", attack.shared_from_this(), secondary_attack)) {
-		if(ability.teacher_loc == a) {
-			continue;
-		}
-
-		if(ability.teacher_loc == b) {
-			continue;
+	const std::vector<std::string> leader_tags{"leadership", "damage", "chance_to_hit", "berserk", "swarm", "drains", "heal_on_hit", "plague", "slow", "petrifies", "firststrike", "poison"};
+	if(!leader_tags.empty()) {
```

Comment:
In what situation could `leader_tags` be empty here?

Reply:
if an ability is other type of what leadership and other in leader_tags like heals or hides

Reply:
Unless I'm completely missing something here, isn't `leader_tags` just a const vector that's always initialized to the same list of values?